alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/CycleService.java

```diff
@@ -0,0 +1,134 @@
+package eu.xenit.alfresco.processor.service;
+
+import eu.xenit.alfresco.processor.model.Cycle;
+import lombok.AllArgsConstructor;
+import org.alfresco.repo.solr.Transaction;
+import org.alfresco.repo.transaction.RetryingTransactionHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@AllArgsConstructor
+public class CycleService {
+    private static final Logger logger = LoggerFactory.getLogger(CycleService.class);
+
+    protected final RetryingTransactionHelper retryingTransactionHelper;
+
+    public void execute(HealthProcessorConfiguration configurationService) {
+        Cycle cycle = createCycle( configurationService);
+        AtomicBoolean continueCycle = new AtomicBoolean();
+        run(configurationService, cycle, continueCycle);
+        while(configurationService.isEnabled()
+                && !configurationService.isRunOnce()
+                && continueCycle.get()) {
+            run(configurationService, cycle, continueCycle);
+        }
+    }
+
+    Cycle createCycle(HealthProcessorConfiguration configurationService) {
+        return new Cycle(
+                configurationService.getTransactionLimit(),
+                configurationService.getFirstTransaction(),
+                configurationService.getTimeIncrementSeconds(),
+                configurationService.getFirstCommitTime(),
+                configurationService.getFirstTransaction(),
+                configurationService.getFirstCommitTime()
+        );
+    }
+
+    void run(HealthProcessorConfiguration configurationService, Cycle cycle, AtomicBoolean continueCycle) {
+        start(cycle);
+
+        AtomicBoolean reachedMaxTx = new AtomicBoolean(false);
+        retryingTransactionHelper.doInTransaction(() -> {
+            reachedMaxTx.set(reachedLastTx(cycle.getTransactionId()));
+            return null;
+        },false, true);
+
+        if(reachedMaxTx.get()) {
+            try {
+                logger.error("Max transaction reached, entering idle state");
+                Thread.sleep(configurationService.getTimeIncrementSeconds() * 1000);
+            } catch (InterruptedException e) {
+                logger.error("Idling has failed, aborting...", e);
+                continueCycle.set(false);
+            }
+        }
+    }
+
+    void start(Cycle cycle) {
+        int txnLimit = cycle.getTxnLimit();
+        int timeIncrementSeconds = cycle.getTimeIncrementSeconds();
+        long firstCommitTime = cycle.getFirstCommitTime();
+
+        logger.debug("Tracking changes ... Start commit time: {}", LocalDateTime
```

Comment:
It would be nice if we can offload all the time format logic to some dedicated helpers and focus on what `start()` means
It is tough to follow what is going on between the epochs, time in millisconds, time in seconds